Bugfixes re. cigar lines, alignment length, etc

muffato opened the pull request:
## Description

We have faced several issues in the EPO and Pecan pipelines:
- ENSCOMPARASW-3913: some ancestral genomic_aligns are longer than the leaves
- ENSCOMPARASW-2054: some cigar lines don't match the sequence (i.e. have a wrong number of `M`s)
- ENSCOMPARASW-4072: some trees are not binary

plus other tickets that have been closed (dismissed) in the past.

## Overview of changes

This PR is a collection of many related changes, which I have reordered in three sections, see below.

### Misc. changes about cigar lines

[→ diff](
https://github.com/Ensembl/ensembl-compara/compare/99796663d24e4d14c9d8109ca79a597030d5ab46..886836a23e4e9250b981fcdbdb8a0427da462d9d)

- Don't store `NULL` as a cigar line
- Don't count `I` elements when computing the alignment length
- New method to compute the sequence length from a cigar line
- Small changes in `assert_valid_cigar`

### A simplification of `RunnableDB::Ortheus`'s Fasta parser

Commit: ccbbe82a8831990fd6f024dda8b74a32670fa124

The Runnable was parsing the Fasta file line by line, assembling the sequence, etc, at the same time it was building the block. I find it much simpler to call a parser and get the sequences one by one. It makes the code more straightforward and easier to read / maintain.

I chose the ensembl-io parser because it reads and returns the sequences one at a time, as opposite to the BioPerl parser, which loads the entire file in memory first.

### The actual bugfixes and safeguards against future bugs

[→ diff](https://github.com/Ensembl/ensembl-compara/compare/ccbbe82a8831990fd6f024dda8b74a32670fa124..bugfix/epo_cigar_line)

- Quite a big diff because Ortheus, Pecan, and LowCoverageAlignment (or whatever new name it will get) now have extra code to check that the alignment is sane:
  - the cigar-line M elements match the sequence length
  - all alignment strings (from the cigar lines) are equal
  - (almost) all trees are binary. There is one case when a ternary node is allowed

  The checks run in the `write_output` transaction, right after storing the alignments, and fetch what has just been stored. If anything is wrong they will backup the temp directory, and create additional reports about the block, the tree, etc, to facilitate debugging. The transaction fails and the database is rolled back, and remains clean.
  I have run these pipelines on _all_ clades (except the pig breeds) and they're all clean run - no such errors. I would still keep the code around for a few releases just in case an error happens again ? I know it's a lot of code... I can remove these commits if you prefer.
  We know that in some cases the error magically goes away upon rerunning, so we may not even notice that there's been a failure (if `retry_count` is > 0)... I can mark the error as non-transient to force eHive to switch the job to `FAILED` if you prefer ?
- To help debugging I have added a script to fake Ortheus' output. It will reformat / return the output of a previous run of Ortheus (which you can find in the backup made by the pipeline upon error). To set it up in your pipeline do this:
  ```bash
  $ tweak_pipeline.pl -url $EHIVE_URL -tweak "analysis[ortheus%].param[ortheus_bin_dir]=$ENSEMBL_CVS_ROOT_DIR/ensembl-compara/scripts/debug"
  ```
  For that reason, I made `RunnableDB::Ortheus` list the sequences always in the same order
- All three analyses now cleanup the worker's temp directory _before_ starting writing data. Previously it was done at the end of `write_output` meaning that if anything went wrong during the job execution, the directory was *not* cleaned. I believe this makes the `tmp_work_dir` functionality of Pecan redundant (which was added exactly because we noticed that there were some leftover files), but I have left it here for now.
- `GenomicAlignTree::restrict_between_alignment_positions` was leaving some internal nodes unrestricted (full length) when they would have become gap-only. This caused the length discrepancy errors. It now restricts _all_ nodes.
- `GenomicAlignTree::restrict_between_alignment_positions` could return a tree with `ancestral_sequences` genomic-aligns as leaves. This may happen if Ortheus thinks that the LCA of two gaps is a non-gap character (true story). These nodes are now removed too and the tree minimised, so that the output tree complies to these rules:
  - binary
  - all nodes are restricted
  - all internal nodes are `ancestral_sequences`
  - all leaves are extant species
  - no leaf is gap-only
- There is a unit-test for  `GenomicAlignTree::restrict_between_alignment_positions`. Given a template tree, I put gaps in various nodes, restrict the tree, and check that the restriction follows the rules above.
- `GenomicAlignTree::annotate_node_type` (only used by the REST API) has been updated to cater for the only case where a node can be ternary. By the way, this only happens in EPO-Extension, when the tree has exactly 3 sequences, because `semphy` builds unrooted trees.
- Tiny but important bugfix in `RunnableDB::Ortheus::detect_pecan_ortheus_errors`. The parsing of the error was expecting some trailing lines that are printed by Ortheus, but these lines are not there when Pecan is run directly. This can be simply recovered by checking whether there a stack trace was being recorded, and register it. This means that the Pecan Runnable (which uses this function) could not automatically process the errors and dataflow to the next analysis / discard the sequences. Although e101 was more complex than e103 because it had a lot more species, I believe that it played a role in making the e101 run of Pecan so difficult and yield a ~20-25% lower coverage on average.
- Added another Pecan exception: `java.lang.ArrayIndexOutOfBoundsException`
- `RunnableDB::Ortheus::split_if_empty_ancestral_seq` now knows that internal nodes too can be gap-only, and splits the tree around them (in up to 3 pieces). Before it could only do that for the root node (recursively). It also shrinks the alignment by removing gap-only columns. This code is now called when the block is large (>1Mbp) and the Runnable needs to split it in chunks.
- `RunnableDB::Ortheus` now does the infamous `sleep(30)` when it detects some missing / empty output files. I had those cases in my tests: MEMLIMIT that are not recorded on the right jobs.

## Testing

As I've said I have rerun the pipelines on almost all clades, and there is a unit test for an important function of `GenomicAlignTree`

## Notes

The sections could perhaps be different pull-requests, if you find it clearer ?

muffato commented:
I guess the conflict is just that the file has been renamed (+ some internal comments / names changed), cf [#249](https://github.com/Ensembl/ensembl-compara/pull/249/files#diff-6fb2e768cb54a6ea0e9ef5a90ff7928d32d8b780149cbf245b8298b0ca525133). I'll fix the conflict after the review, when merging this PR

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/MercatorPecan/Pecan.pm:
```diff
@@ -272,16 +278,84 @@ sub _write_output {
+        $lengths{$aln_length}++;
+    }
+    if (scalar(keys %lengths) > 1) {
+        $self->_backup_data_and_throw($gab, "Multiple alignment lengths: ". stringify(\%lengths));
```
```suggestion
        $self->_backup_data_and_throw($gab, "Multiple alignment lengths: " . stringify(\%lengths));
```

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/EpoLowCoverage/LowCoverageGenomeAlignment.pm:
```diff
@@ -225,17 +232,119 @@ sub _write_output {
+    throw("Error: $err\nData copied to $target_dir");
+}
+
+sub _print_regions {
+    my ($self, $filename) = @_;
+    open(my $fh, '>', $filename);
+    foreach my $region (@{$self->param('genomic_aligns')}) {
+        print $fh join("\t",
+            $region->genome_db->name,
+            $region->genome_db->dbID,
+            $region->dnafrag->name,
+            $region->dnafrag_id,
+            $region->dnafrag->length,
+            $region->dnafrag_start,
+            $region->dnafrag_end,
+            $region->dnafrag_strand,
+        ), "\n";
+    }
+    close($fh);
+}
```
This is identical to `_print_regions` in `Pecan.pm` (replacing `genomic_aligns` by `dnafrag_regions`). Maybe it could be worth making it a utils function of sorts?

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/EpoLowCoverage/LowCoverageGenomeAlignment.pm:
```diff
@@ -225,17 +232,119 @@ sub _write_output {
+        }
+    }
+    if (scalar(keys %lengths) > 1) {
+        $self->_backup_data_and_throw($gat, "Multiple alignment lengths: ". stringify(\%lengths));
```
```suggestion
        $self->_backup_data_and_throw($gat, "Multiple alignment lengths: " . stringify(\%lengths));
```

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/EpoLowCoverage/LowCoverageGenomeAlignment.pm:
```diff
@@ -225,17 +232,119 @@ sub _write_output {
+    }
+}
+
+sub _assert_cigar_lines_in_block {
```
I have considered moving this method in a generic form to `Utils/Cigars.pm` since you are using it in three different runnables, but it may be too complicated based on the many tailored bits (although the core is the same). Leaving it here just in case :)

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/EpoLowCoverage/LowCoverageGenomeAlignment.pm:
```diff
@@ -225,17 +232,119 @@ sub _write_output {
+            my $cigar_length = Bio::EnsEMBL::Compara::Utils::Cigars::sequence_length_from_cigar($genomic_align->cigar_line);
+            my $region_length = $genomic_align->dnafrag_end - $genomic_align->dnafrag_start + 1;
+            if ($cigar_length != $region_length) {
+                $self->_backup_data_and_throw($gat, "cigar_line's sequence length ($cigar_length) doesn't match the region length ($region_length)" );
```
```suggestion
                $self->_backup_data_and_throw($gat, "cigar_line's sequence length ($cigar_length) doesn't match the region length ($region_length)");
```

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/MercatorPecan/Pecan.pm:
```diff
@@ -272,16 +278,84 @@ sub _write_output {
+        my $cigar_length = Bio::EnsEMBL::Compara::Utils::Cigars::sequence_length_from_cigar($genomic_align->cigar_line);
+        my $region_length = $genomic_align->dnafrag_end - $genomic_align->dnafrag_start + 1;
+        if ($cigar_length != $region_length) {
+            $self->_backup_data_and_throw($gab, "cigar_line's sequence length ($cigar_length) doesn't match the region length ($region_length)" );
```
```suggestion
            $self->_backup_data_and_throw($gab, "cigar_line's sequence length ($cigar_length) doesn't match the region length ($region_length)");
```

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/EpoLowCoverage/LowCoverageGenomeAlignment.pm:
```diff
@@ -225,17 +232,119 @@ sub _write_output {
+    my ($self, $gat) = @_;
+
+    if ($gat->get_child_count == 3) {
+        # The only ternary nodes allowed are "root" nodes of unroot trees
```
```suggestion
        # The only ternary nodes allowed are "root" nodes of unrooted trees
```

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/EpoLowCoverage/LowCoverageGenomeAlignment.pm:
```diff
@@ -225,17 +232,119 @@ sub _write_output {
+        # of 3 sequences
+        foreach my $child (@{$gat->children}) {
+            unless ($child->is_leaf) {
+                $self->_backup_data_and_throw($gat, sprintf("Node %s is the child of a ternay node, but node a leaf of the tree", $child->name));
```
```suggestion
                $self->_backup_data_and_throw($gat, sprintf("Node %s is the child of a ternay node, but not a leaf of the tree", $child->name));
```

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/Ortheus.pm:
```diff
@@ -316,33 +331,122 @@ sub _write_output {
   	       }
   	       my $new_gat = $genomic_align_tree->restrict_between_alignment_positions($start, $end, "skip_empty_GenomicAligns");
-	       push @$split_trees, $new_gat if scalar(@{$new_gat->get_all_leaves}) >= 2;
+	       # Some ancestral genomic_aligns may have become full-gaps,
```
```suggestion
	       # Some ancestral genomic_aligns may have become all gaps,
```

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/Ortheus.pm:
```diff
@@ -316,33 +331,122 @@ sub _write_output {
+            my $cigar_length = Bio::EnsEMBL::Compara::Utils::Cigars::sequence_length_from_cigar($genomic_align->cigar_line);
+            my $region_length = $genomic_align->dnafrag_end - $genomic_align->dnafrag_start + 1;
+            if ($cigar_length != $region_length) {
+                $self->_backup_data_and_throw($gat, "cigar_line's sequence length ($cigar_length) doesn't match the region length ($region_length)" );
```
```suggestion
                $self->_backup_data_and_throw($gat, "cigar_line's sequence length ($cigar_length) doesn't match the region length ($region_length)");
```

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/Ortheus.pm:
```diff
@@ -316,33 +331,122 @@ sub _write_output {
+        }
+    }
+    if (scalar(keys %lengths) > 1) {
+        $self->_backup_data_and_throw($gat, "Multiple alignment lengths: ". stringify(\%lengths));
```
```suggestion
        $self->_backup_data_and_throw($gat, "Multiple alignment lengths: " . stringify(\%lengths));
```

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/Ortheus.pm:
```diff
@@ -316,33 +331,122 @@ sub _write_output {
+    }
+}
+
+sub _backup_data_and_throw {
```
Having these three methods repeated almost to a dot in three runnables seems like a good reason to refactor... Would it be worth it?

carlacummins replied:
Agree, but we may be out of time on this one - might have to do it ourselves @JAlvarezJarreta 😉

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/Ortheus.pm:
```diff
@@ -438,63 +540,53 @@ sub parse_results {
+        my $name = $parser->getHeader;
+        my $seq  = $parser->getSequence;
 
+		    my $this_genomic_align = new Bio::EnsEMBL::Compara::GenomicAlign;
 		    $this_genomic_align->aligned_sequence($seq);
 		    $this_genomic_align_block->add_GenomicAlign($this_genomic_align);
```
```suggestion
        my $this_genomic_align = new Bio::EnsEMBL::Compara::GenomicAlign;
        $this_genomic_align->aligned_sequence($seq);
        $this_genomic_align_block->add_GenomicAlign($this_genomic_align);
```
I know those two lines "are not yours" but I think it may be worth for the sake of readability to change them.

muffato replied:
Regarding the indentation of this file, it is best seen with tab=8 spaces, but they do indentation levels of 4, so (1) 4 spaces, (2) 1 tab, (3) 1 tab and 4 spaces, (4) 2 tabs, etc :roll_eyes:
When changing a line, I have kept the original indentation, and when adding a line, I have matched the indentation of the surrounding lines. If we're fine with resetting the indentation to 4 spaces and no tabs (I'm fine with it), that would edit 8 lines, and I would rather go back to my own commits and edit them so that they use the correct indentation right from the start

JAlvarezJarreta replied:
I would vote for the change if you don't mind, since I consider it an improvement. But will wait for the vote from the other reviewer.

carlacummins replied:
I'm not fussy

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/Ortheus.pm:
```diff
@@ -438,63 +540,53 @@ sub parse_results {
-	    }
-	    my $header = shift(@$ids);
-	    $this_genomic_align = new Bio::EnsEMBL::Compara::GenomicAlign;
+		    my $genomic_align_group = new Bio::EnsEMBL::Compara::GenomicAlignGroup;
+		    $genomic_align_group->add_GenomicAlign($this_genomic_align);
```
```suggestion
        my $genomic_align_group = new Bio::EnsEMBL::Compara::GenomicAlignGroup;
        $genomic_align_group->add_GenomicAlign($this_genomic_align);
```

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/Ortheus.pm:
```diff
@@ -438,63 +540,53 @@ sub parse_results {
 
-	    if (!defined($header)) {
-		print "INTERNAL NODE $name\n" if ($self->debug);
 		my $this_node;
 		foreach my $this_leaf_name (split("_", $name)) {
```
```suggestion
        my $this_node;
        foreach my $this_leaf_name (split("_", $name)) {
```
Same reason as before.

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/Ortheus.pm:
```diff
@@ -438,63 +540,53 @@ sub parse_results {
 		    }
 		}
-		print join("_", map {$_->name} @{$this_node->get_all_leaves}), "\n" if ($self->debug);
+		print "Found node: ", join("_", map {$_->name} @{$this_node->get_all_leaves}), "\n" if ($self->debug);
```
```suggestion
        print "Found node: ", join("_", map {$_->name} @{$this_node->get_all_leaves}), "\n" if ($self->{debug});
```
I don't remember the details, but I remember having an issue with `$self->debug` in the past, but the hash access was fine. Maybe I messed something else :man_shrugging:

muffato replied:
Unexpected as it's not stored in `$self` directly, cf https://github.com/Ensembl/ensembl-hive/blob/version/2.5/modules/Bio/EnsEMBL/Hive/Process.pm#L641-L654

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/Ortheus.pm:
```diff
@@ -438,63 +540,53 @@ sub parse_results {
+        my $header = shift(@$ids);
+        if (not defined $header) {
+
+		print "INTERNAL NODE $name\n" if ($self->debug);
```
```suggestion
		print "INTERNAL NODE $name\n" if ($self->{debug});
```
Related with previous comment.

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/Ortheus.pm:
```diff
@@ -438,63 +540,53 @@ sub parse_results {
-			$this_node = $tree->find_node_by_name($this_leaf_name);
+			$this_node = $this_leaf;
 		    }
 		}
```
```suggestion
        }
```

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/Ortheus.pm:
```diff
@@ -438,63 +540,53 @@ sub parse_results {
-		print join("_", map {$_->name} @{$this_node->get_all_leaves}), "\n" if ($self->debug);
+		print "Found node: ", join("_", map {$_->name} @{$this_node->get_all_leaves}), "\n" if ($self->debug);
+
+		$this_node->cast('Bio::EnsEMBL::Compara::GenomicAlignTree');
+		$this_node->genomic_align_group($genomic_align_group);
```
```suggestion
        $this_node->cast('Bio::EnsEMBL::Compara::GenomicAlignTree');
        $this_node->genomic_align_group($genomic_align_group);
```

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/Ortheus.pm:
```diff
@@ -510,56 +602,54 @@ sub parse_results {
-											#-genomic_align_array => [$this_genomic_align],
-										        -type => "epo");
-		    $genomic_align_group->add_GenomicAlign($this_genomic_align);
+		    print "store gag2 $this_node\n" if $self->debug;
```
```suggestion
		    print "store gag2 $this_node\n" if $self->{debug};
```
(same as before)

JAlvarezJarreta commented on modules/Bio/EnsEMBL/Compara/RunnableDB/Ortheus.pm:
```diff
@@ -510,56 +602,54 @@ sub parse_results {
+		    my $original_sequence = $seq;
+		    $original_sequence =~ s/-//g;
+		    if (length($original_sequence) != $region_length) {
+			throw("Length mismatch: $region_length from the coordinates, ".length($original_sequence)." from the aligned string");
```
```suggestion
			throw("Length mismatch: $region_length from the coordinates, " . length($original_sequence) . " from the aligned string\n");
```

JAlvarezJarreta reviewed (approved):
Quite an update, but feels easier to follow and understand now. Good job :+1:
Only a few minor comments and two refactoring suggestions.

muffato commented on modules/t/genomicAlignTree.t:
```diff
@@ -0,0 +1,130 @@
+    my $rtree = $tree->restrict_between_alignment_positions(@restrict_pos, 'skip_empty_GenomicAligns');
+    my %expected_gaps = map {$_ => 1} @$expect_gaps;
+    subtest join('', @$set_gaps, (scalar(@$expect_gaps) ? ('/', @$expect_gaps) : ())) => sub {
+        $tree->print;
+        $rtree->print;
```
Just for debugging, yes

muffato commented on travisci/python-linter_harness.sh:
```diff
@@ -24,13 +24,13 @@ export MYPYPATH=$MYPYPATH:$PWD/src/python/lib
 PYLINT_OUTPUT_FILE=$(mktemp)
 PYLINT_ERRORS=$(mktemp)
-find "${PYTHON_SOURCE_LOCATIONS[@]}" -type f -name "*.py" -print0 | xargs -0 pylint --rcfile pylintrc --verbose | tee "$PYLINT_OUTPUT_FILE"
+find "${PYTHON_SOURCE_LOCATIONS[@]}" -type f -name "*.py" \! -name "Ortheus.py" -print0 | xargs -0 pylint --rcfile pylintrc --verbose | tee "$PYLINT_OUTPUT_FILE"
```
Because it's Python 2. I could get pylint to run on it, but I need a different installation of pylint, so maybe another Travis stage ? I thought it was not worth it because it's the only Python 2 script

muffato commented on travisci/python-linter_harness.sh:
```diff
@@ -24,13 +24,13 @@ export MYPYPATH=$MYPYPATH:$PWD/src/python/lib
 rm "$PYLINT_OUTPUT_FILE" "$PYLINT_ERRORS"
 
-find "${PYTHON_SOURCE_LOCATIONS[@]}" -type f -name "*.py" -print0 | xargs -0 mypy --config-file mypy.ini --namespace-packages
+find "${PYTHON_SOURCE_LOCATIONS[@]}" -type f -name "*.py" \! -name "Ortheus.py" -print0 | xargs -0 mypy --config-file mypy.ini --namespace-packages
```
`mypy` works on Python 2 scripts with the `--py2` option. `Ortheus.py` has to be excluded from this command-line anyway, but I could add another line to run `mypy` on it ?

JAlvarezJarreta replied:
I don't think it is worth it since this script is not meant to be touched and we won't produce any more Python 2 code.

muffato commented on scripts/debug/Ortheus.py:
```diff
@@ -0,0 +1,61 @@
+pid = 85677
+
+# This is where you have saved Ortheus' output from a previous run
+ref_fasta_dir = '/path/to/worker_muffato_mammals_epo_with4x_101.95260'
```
I wonder why it was `4x`, not `2x` :thinking:

muffato commented on modules/Bio/EnsEMBL/Compara/GenomicAlignTree.pm:
```diff
@@ -1333,28 +1363,30 @@ sub annotate_node_type {
+                if ($duplications->{$name}) {
+                    # Found a duplication. Stop all for loops of this node
+                    $self->node_type("duplication");
+                    last OUTER;
```
If you've got an idea to make this look better, I'm open to suggestions :)

JAlvarezJarreta replied:
I'm ok with it, but couldn't resist the comment ;)

muffato commented on modules/Bio/EnsEMBL/Compara/GenomicAlignTree.pm:
```diff
@@ -655,9 +660,13 @@ sub restrict_between_alignment_positions {
+  my $all_nodes = $genomic_align_tree->get_all_nodes;
+  my %was_leaf = map {$_ => $_->is_leaf} @$all_nodes;
+
+  # Depth-first, children-first, traversal of the tree
```
Absolutely
```suggestion
  # Postorder traversal of the tree
```

muffato commented on scripts/debug/Ortheus.py:
```diff
@@ -0,0 +1,61 @@
+import sys
+
+# Set here the expected inputs, which you can find in the Ortheus command line
+species_tree = '(((((((327:0.0585842,457:0.0598399):0.0310988,(180:0.00823412,179:0.0124659):0.09194331):0.00755855,(((334:0.0638217,214:0.064933836):0.0036732,190:0.0702538):0.0105362,((((174:0.01004235,134:0.010315324):0.0109081,212:0.0204159):0.016895,213:0.0365203):0.0231325,155:0.0629363):0.016747061):0.02896091):0.00360012,108:0.1033556):0.000861311,((((((((221:0.00217461,210:0.00336539):0. ...
+species_list = "60 98 98 108 108 108 134 134 135 135 135 147 153 174 179 180 190 190 198 211 212 222 237 285 285 285 317 327 327 334 334 361 372 372 372 379 379 379 387 394 396 407 407 407 416 429 434 434 434 435 443".split()
```
I'm not too convinced about this suggestion

JAlvarezJarreta replied:
It was to make it "PEP8 compatible", but since this is a "unique" file, I guess we can ignore it.

muffato commented on scripts/debug/Ortheus.py:
```diff
@@ -0,0 +1,61 @@
+        return fh.read()
+
+# Check that we are being called with the correct arguments
+assert sys.argv[-4-len(species_list):-4] == species_list
+assert sys.argv[-4-len(species_list)-2] == species_tree
```
Sure. There are other `sys.argv[XXX]` below that can be explained too :+1:

muffato commented:
Yes, the functions to assert / print the data are all very similar and can be factored out. I didn't do it initially because I wasn't sure if we wanted to keep this code in the Runnables

carlacummins commented on scripts/debug/Ortheus.py:
```diff
@@ -0,0 +1,71 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Script to mimic Ortheus and return the output of a previous run"""
```
Could you add the instructions about how to use this (with `tweak_pipeline.pl` as you have in the PR description) in here somewhere?

carlacummins reviewed (commented):
> We know that in some cases the error magically goes away upon rerunning, so we may not even notice that there's been a failure (if retry_count is > 0)... I can mark the error as non-transient to force eHive to switch the job to FAILED if you prefer ?

I think it's fine as it is. The worst thing that happens is that we back up some transient data that we don't actually need to inspect in the end, right? I can only see that being an issue if a *lot* of jobs are following this pattern (which I don't think they are).

muffato commented:
I have rebased the branch onto `release/104` to fix the conflict with `LowCoverageAlignment.pm` and I have cleaned up the indentation in `Ortheus.pm`.
Note that the commit IDs listed in the description of this PR won't match the actual commits any more.
